File: be/src/main/java/nl/zzpmatcher/userLogon/security/SecurityBeans.java
package nl.zzpmatcher.userLogon.security;

import nl.zzpmatcher.userLogon.business.User;
import nl.zzpmatcher.userLogon.business.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Collection;
import java.util.stream.Collectors;

@Component
public class SecurityBeans {

    private UserRepository userRepository;

    @Autowired
    public SecurityBeans(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Bean
    public UserDetailsService userDetailsService() {

        return new UserDetailsService() {
            @Override
            public UserDetails loadUserByUsername(String s) throws UsernameNotFoundException {

                final User user = userRepository.findUserByEmailaddress(s);

                return new UserDetails() {
                    @Override
                    public Collection<? extends GrantedAuthority> getAuthorities() {

                        return Arrays
                                .stream(getIndividualRoles())
                                .map(this::mapToGrantedAuthority)
                                .collect(Collectors.toList());
                    }

                    private GrantedAuthority mapToGrantedAuthority(String role) {
                        return () -> role;
                    }

                    private String[] getIndividualRoles() {
                        return user.getRoles().split(",");
                    }

                    @Override
                    public String getPassword() {
                        return user != null ? user.getPassword() : null;
                    }

                    @Override
                    public String getUsername() {
                        return user.getEmailaddress();
                    }

                    @Override
                    public boolean isAccountNonExpired() {
                        return true;
                    }

                    @Override
                    public boolean isAccountNonLocked() {
                        return true;
                    }

                    @Override
                    public boolean isCredentialsNonExpired() {
                        return true;
                    }

                    @Override
                    public boolean isEnabled() {
                        return true;
                    }
                };
            }
        };
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

}

File: be/src/main/java/nl/zzpmatcher/userlogon/security/SecurityBeans.kt
package nl.zzpmatcher.userlogon.security

import nl.zzpmatcher.userlogon.business.User
import nl.zzpmatcher.userlogon.business.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component

import java.util.Arrays
import java.util.stream.Collectors
import kotlin.streams.toList

@Component
class SecurityBeans @Autowired
constructor(private val userRepository: UserRepository) {

    @Bean
    fun userDetailsService(): UserDetailsService {

        return UserDetailsService { s ->
            val user = userRepository.findUserByEmailaddress(s)

            object : UserDetails {
                override fun getAuthorities(): Collection<GrantedAuthority> {

                    return Arrays
                            .stream(individualRoles)
                            .map { GrantedAuthority({it}) }
                            .toList()
                }

//                private fun mapToGrantedAuthority(role: String): GrantedAuthority {
//                    return { role }
//                }

                private val individualRoles: Array<String>
                    get() = user.roles!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                override fun getPassword(): String? {
                    return user?.password
                }

                override fun getUsername(): String? {
                    return user.emailaddress
                }

                override fun isAccountNonExpired(): Boolean {
                    return true
                }

                override fun isAccountNonLocked(): Boolean {
                    return true
                }

                override fun isCredentialsNonExpired(): Boolean {
                    return true
                }

                override fun isEnabled(): Boolean {
                    return true
                }
            }
        }
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }

}
